Source/MQTTnet/Client/MqttClient.cs

```diff
@@ -561,12 +561,16 @@ async Task TryProcessReceivedPacketAsync(MqttBasePacket packet, CancellationToke
                 {
                     await SendAsync(new MqttPingRespPacket(), cancellationToken).ConfigureAwait(false);
                 }
-                else if (packet is MqttDisconnectPacket)
+                else if (packet is MqttDisconnectPacket disc)
                 {
                     // Also dispatch disconnect to waiting threads to generate a proper exception.
                     _packetDispatcher.Dispatch(packet);
 
-                    await DisconnectAsync(null, cancellationToken).ConfigureAwait(false);
+                    await DisconnectAsync(new MqttClientDisconnectOptions() 
+                    {
+                        // todo conversion
+                        ReasonCode = disc.ReasonCode
```

Comment:
Is this only the conversion between the client  API and the internal protocol enum? If yes I usually cast the enum as int and then to the new enum (if positions ans numbers are matching). The reason is that the client may have less enum values because other code is handling this or it is dangerous if the user is able to set some "wrong" values which do not follow the internal logic etc.